### DISCUSSION TITLE
Explicitly ordered database migrations

### DISCUSSION THREAD

**Discussion by dafydd-t** [grafana/grafana#41320] [Platform]
Currently in Grafana, functions which perform migrations are added to [this list of migrations](https://github.com/grafana/grafana/blob/7422189a88caf28f3c29b3416cda689a656eb232/pkg/services/sqlstore/migrations/migrations.go#L20). This list of functions is ordered, but multiple migrations can happen within each function ([example](https://github.com/grafana/grafana/blob/main/pkg/services/sqlstore/migrations/alert_mig.go#L7)). If an additional migration is added to a function, this new migration will be performed before any migrations in functions listed after the changed function. This means the order of existing migrations in Grafana is not guaranteed across time.

We also have a migration that is only run based on certain environment variables ([example](https://github.com/grafana/grafana/blob/7422189a88caf28f3c29b3416cda689a656eb232/pkg/services/sqlstore/migrations/ualert/ualert.go#L54-L80)). This is the only example of a migration (that I have encountered - [I'm not considering this migration as the function is empty](https://github.com/grafana/grafana/blob/7422189a88caf28f3c29b3416cda689a656eb232/pkg/services/sqlstore/migrations/migrations.go#L51-L53)) that has this behaviour, and it happens to be a data migration, rather than a DDL migration.

These kinds of behaviours have caused issues when operating Grafana at scale. The issues occur when upgrading / downgrading instances. ([example](https://docs.google.com/document/d/16mLC2IavZVfFzlzrchZXeNsLxjmGv4wViTepL5InA3w/edit#heading=h.qbg8mtn8fn1g))

### What I would like to see

- I would like to see Grafana use an append-only list of migrations, where any new migration must be run after all other migrations have run. 
- Data can still be modified when feature toggles are enabled/disabled
- Logic within CI to defend these characteristics

### How I think we can achieve this

_Before listing suggestions, I'd like to add the caveat that there will be many others who have a better understanding of the Grafana codebase, especially around migrations, who might be able to disagree with these suggestions or offer better approaches. I'd love to hear from those people._

**Every individual migration should have a sequence number, and any new migration should have the highest sequence number of all migrations.**

These sequence numbers do not necessarily need to be persisted in the migration log (therefore no change to the DB schema would be required). Some structure within the migrator could ensure that migrations are run in sequence order. Add logic to error if a duplicate sequence number is added. 

An alternative to this approach is to force all migrations to a single top-level list. This could make it more easy to see the order in which migrations are run, but it may require a larger change to the code structure.

**DDL migrations, even for features behind toggles, should not be conditional on the toggle. Conditional data migrations should be handled by the feature, not the main migrations system**

This means for a new feature, the DDL changes required to enable that feature would be always visible.

For a concrete example, I would suggest that [these data migrations](https://github.com/grafana/grafana/blob/7422189a88caf28f3c29b3416cda689a656eb232/pkg/services/sqlstore/migrations/ualert/ualert.go#L54-L80) would be run by the ualert service, rather than added to the migrator.

The benefit of this is that the migration log table would only grow, never changed in-place.

**Test to defend the final output of the migrations**

This is dependent on the migration logs table only containing changes that are *not* dependent on env variables. One approach here is to use a ["golden file"](https://ieftimov.com/post/testing-in-go-golden-files/) to define exactly what the output of the migrations log table should be after the migrations are run. Any new migration would have to be added to this file. Some test in CI could compare the result of a run of the migrations against this file. This would guarantee that the order of migrations does not change by accident, and we can easily defend the append-only rule in PR by looking at this golden file.

cc @sdboyer for reference
cc @papagian because I think you'll have some good opinions on this
cc @joanlopez as I think you've been affected by some of these issues

Relevant internal docs:
[[1]](https://docs.google.com/document/d/1JPdi7vWC1zYUbCq5hvVc7kbBaRe-gIi_AUH1j6XrZR8/edit#heading=h.uwvn7wgpa2t7)
[[2]](https://docs.google.com/document/d/1jF9eOhyQvAkRRfkzIUL21ToIGvf-VVAAw4Q--dsBVhg/edit#heading=h.9blr5foqfzwb)

**Comment by joanlopez** [comment]
Hey @dafydd-t,

Thanks a lot for taking care of this, as I mentioned previously, that's something we really need to get done sooner than later, not a nice to have. So, I'll try to help as much as possible and support this initiative! ❤️‍🔥 

That being said, I leave you here some feedback on your initial description:

> We also have a migration that is only run based on certain environment variables (example). This is the only example of a migration (that I have encountered - I'm not considering this migration as the function is empty) that has this behaviour, and it happens to be a data migration, rather than a DDL migration.

I've seen at least one more in Enterprise behind a FT (maybe the FT has been removed now, but similar new ones could be potentially be added). I that case it was not data migration that can go back and forth, but a migration that it's only executed if that FT is turned on while running migrations.

> Every individual migration should have a sequence number, and any new migration should have the highest sequence number of all migrations.
> An alternative to this approach is to force all migrations to a single top-level list. This could make it more easy to see the order in which migrations are run, but it may require a larger change to the code structure.

Something I'd consider before falling into this approach is how to handle order on Git conflicts. For example, using a list (or even a simple number sequence) could be much easier than other strategies (like timestamps or other time-based values).

> DDL migrations, even for features behind toggles, should not be conditional on the toggle. Conditional data migrations should be handled by the feature, not the main migrations system.
> This means for a new feature, the DDL changes required to enable that feature would be always visible.
>For a concrete example, I would suggest that these data migrations would be run by the ualert service, rather than added to the migrator.
> The benefit of this is that the migration log table would only grow, never changed in-place.

I love it! ❤️ That's precisely what we discussed with @undef1nd 

> Test to defend the final output of the migrations

No strong opinions on that, but I agree somehow doing it would be ideal (and probably a must).

Thanks!!! 😊

**Comment by joanlopez** [comment]
Also related: https://github.com/grafana/grafana-enterprise/discussions/2557

**Comment by dafydd-t** [comment]
[Design doc on this topic](https://docs.google.com/document/d/1RuOFHTBQ3Zh6SIERf3qTVoR-Gy4dQ8izwkUSjOpsjBA/edit#). Proposes immutable migration functions and a method to enforce them

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
